### ISSUE TITLE
Best Practices for Async Map/Element Management

### ISSUE THREAD

**Issue by leepeterson** [bozdoz/wp-plugin-leaflet-map#48]
Howdy. Absolutely love what you've created, and pleased even more with your continued development and the improvements you've been making. So, thanks! and kudos, good sir.

Wanted to ask your opinion on handling asynchronous loading of new maps, markers, GeoJSON, polylines, etc. given how you've built the underlying components. Perhaps the best way to phrase this request is to provide an example or two.

1. Given multiple maps generated on a single page, subsequent maps living within a tabbed interface (ie. map 1 is visible; maps 2-whatever contained within a tabpanel, visible only when another tab is activated): what would be the best way to handle this using your plugin? Similarly with lazy loaded DOM elements, this plugin's scripts fire once on window load, then never again, and these additional map elements aren't initialized.

2. Given multiple sets of elements for a single map (markers, lines, GPX/KML, etc.), how could I make best use of your plugin to initialize additional sets of elements via, say, a sidebar which sits adjacent to a map?

3. I just saw your excellent efforts with rewriting the plugin using classes, making use of various PSR-X class types... this is by far the most exciting change I've seen in your plugin... and it appears that you're planning to allow modifications using WordPress-style hooks (actions, filters). Would you consider `class.shortcode.php` ready for external use now? Think I could extend the `Leaflet_Shortcode` abstract class to begin extending the functionality of this plugin to avoid modifying its core files?

Hope that wasn't too wordy ;) Simple yes/no and maybe a link or reference is all I'm hoping for. Appreciate your time and efforts! Thank you!

**Comment by bozdoz** [comment]
Let me try to clarify some points for my own sake:

1. I think I understand this to mean that you want to conditionally render maps (say, after a user clicks a tab).  I suppose for this case, perhaps I (or you) could extend the JavaScript portion of the plugin (./scripts/construct-leaflet-map.min.js): alter `this.add`, or `this.init` to conditionally render a map after a different event.

2. You could manipulate a given map with additional custom JavaScript: `WPLeafletMapPlugin.maps` is an array of maps.  So you could add to these maps after they're rendered: `WPLeafletMapPlugin.maps[0].addLayer( layer );` for example.  I hope that helps.  Not sure of a better way to do that.  I suppose we could create the maps as an Object with a name as the key and the map as the value, and create it like: `[leaflet-map name="moose"]`; and get it via JavaScript like: `WPLeafletMapPlugin.maps['moose']`.  Anyway, that could be easy to accomplish and might be helpful for extending.

3. You can definitely try to extend `class.shortcode.php`.  All of the shortcodes are currently extending from it.  It's not a great implementation of it.  What you could do, perhaps, is add an `add_action` for `leaflet-map-loaded`, then add the shortcode, similar to `add_shortcodes` in `class.leaflet-map.php`.

Let me know if I understood you correctly.  Thanks! :smile:

**Comment by bozdoz** [comment]
If I misunderstood you, open a new issue, thanks!
